src/crypto/dleq.rs

```diff
@@ -0,0 +1,1047 @@
+
+        assert!(ecdsa.verify(&xH_p, &pok_1_message_hash, &pok_1));
+
+        DLEQProof {
+            xG_p,
+            xH_p,
+            c_g,
+            c_h,
+            ring_signatures,
+            pok_0,
+            pok_1,
+        }
+    }
+
+    pub(crate) fn verify(&self) -> Result<(), crypto::Error> {
+        assert_eq!(252, self.c_g.len());
+        assert_eq!(252, self.c_h.len());
+        assert_eq!(252, self.ring_signatures.len());
+
+        // Commitments
+        let commitment_agg_ed25519 = self.c_g.iter().sum();
+
+        if !(self.xG_p == commitment_agg_ed25519) {
+            return Err(crypto::Error::InvalidPedersenCommitment);
+        }
+
+        let commitment_agg_secp256k1 = self
+            .c_h
+            .iter()
+            .fold(secp256k1Point::zero(), |acc, bit_commitment| {
+                g!(acc + bit_commitment).mark::<Normal>()
+            });
+
+        if !(self.xH_p == commitment_agg_secp256k1) {
+            return Err(crypto::Error::InvalidPedersenCommitment);
+        }
+
+        // Ring signatures
+        let valid_ring_signatures = self
+            .c_g
+            .clone()
+            .iter()
+            .enumerate()
+            .zip(self.c_h.clone())
+            .zip(self.ring_signatures.clone())
+            .all(|(((index, c_g_i), c_h_i), ring_sig)| {
+                verify_ring_sig(index, *c_g_i, c_h_i, ring_sig)
+            });
+
+        if !(valid_ring_signatures) {
+            return Err(crypto::Error::InvalidRingSignature);
+        }
+
+        // Proof of Knowledge
+        // ed25519 (edDSA)
+        #[allow(non_snake_case)]
+        let (alpha_G, r) = self.pok_0;
+        let mut challenge_preimage = vec![];
+        challenge_preimage.extend_from_slice(alpha_G.compress().as_bytes());
+        challenge_preimage.extend_from_slice(self.xG_p.compress().as_bytes());
+        let pok_0_message: Vec<u8> = self.c_g.iter().fold(vec![], |mut acc, pc| {
+            acc.extend_from_slice(pc.compress().as_bytes());
+            acc
+        });
+        challenge_preimage.extend(pok_0_message);
+        let challenge = monero::cryptonote::hash::keccak_256(&challenge_preimage);
+
+        if r * G != alpha_G + ed25519Scalar::from_bytes_mod_order(challenge) * self.xG_p {
+            return Err(crypto::Error::InvalidProofOfKnowledge);
+        }
+
+        // secp256k1 (ECDSA)
+        let ecdsa = ecdsa_fun::ECDSA::verify_only();
+        let pok_1_message = self.c_h.iter().fold(vec![], |mut acc, pc| {
+            acc.extend_from_slice(&pc.to_bytes());
+            acc
+        });
+        let pok_1_message_hash: [u8; 32] = sha2::Sha256::digest(pok_1_message.as_slice())
+            .try_into()
+            .unwrap();
+        if !ecdsa_fun::ECDSA::verify(&ecdsa, &self.xH_p, &pok_1_message_hash, &self.pok_1) {
+            return Err(crypto::Error::InvalidProofOfKnowledge);
+        }
+
+        // Everything ok
+        Ok(())
+    }
+}
+
```

Comment:
Maybe add the convention of test modules:
```rust
#[cfg(test)]
mod tests {
    ...
}
```

Reply:
I'll do this in #144 - otherwise too much rebasing.